rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,473 @@
+1.0 release.
+
+### Stakeholders
+
+Let's first set out to define the different needs that the various stakeholders
+of the Tendermint software have in the context of upgrading:
+
+- Application Developers: want to be able to get as much of the benefits of
+  new features (like state sync and light client) with minimal work or loss to
+  their current infrastructure. Ideally, Tendermint is also somewhat
+  accommodating to application upgrades.
+
+- Node Operators: Primarily they want safety and reliability. This means minimal
+  down-time and intuitive / easy UX when it comes to upgrading.
+
+- Wallets, Block Explorers and other clients: data retrievability which means
+  ensuring that helpful/informative data is always available and that minimal
+  infrastructure is needed to support serving data across the entire lifespan of
+  the chain.
+
+### Protocol Versioning
+
+Tendermint currently has three protocol versions: Block, P2P and App.
+
+The block protocol version would initially be the sole focus for soft upgrades.
+This consists of all block related data structures i.e. `Header`,
+`Vote`, `Validator`, `Commit` (a full list of the exact data structures can be
+found in [data structures](https://github.com/tendermint/spec/blob/master/spec/core/data_structures.md#data-structures).
+Data structures are dictated by the spec thus requiring a Tendermint software
+version to indicate the block version/s it supports. When a new spec release
+makes changes to any of the aforementioned data structures, this will result in
+incrementing the block protocol version.
+
+The app version can already be changed via consensus and is left to the
+discretion of the application to handle. The P2P version would also be left as
+it is, with major revisions requiring a hard upgrade. However, this may be
+revisited in the future.
+
+### Transitions
+
+Networks would be expected to announce upgrades in advance allowing a window for
+node operators to upgrade to a software version of Tendermint that would
+comply with the announced block protocol upgrade. This can be done
+asynchronously with node operators temporarily halting their node, changing to
+the new binary and starting up again. Tendermint could also offer a tool to
+easily swap between binaries thus reducing the amount of down-time that nodes
+experience although this is outside the scope of this RFC.
+
+The actual switch would occur in the same manner as an App Version change, using
+the `EndBlockResponse{}` ABCI call to indicate to Tendermint that a different
+block protocol be used. In the case that a node failed to upgrade, the node
+would gracefully stop and error.
+
+Given the delayed execution model, careful consideration would be needed to
+ensure that the nodes could transition smoothly. It may be that the actual
+upgrade might not take place till height: h + 2.
+
+As we expect upgrades to always increment the block version by one, rather
```

Comment:
Is there a reason to enforce this? I could imagine an upgrade skipping some versions, but maybe the idea is that it will be easier to provide backwards compatibility in the software for 1 rather than 2 or more past versions?

Reply:
Hmm no. I was rather presuming this nature but I agree that it might be beneficial to allow skipping versions (even potentially reverting versions if for example a bug was found)

Reply:
> As an example, imagine that Software Version X supports Block Versions Y and Y+1. When you want to introduce Block Version Y+2, you could release Software Version X+1, which would support Y+1 and Y+2 but could drop support for Block Version Y.

Hmm, the problem with this is that all the heights that were made during Block Version Y are now lost (or can only be retrieved by processes that can still read and validate Block Version Y. This is not a problem for light clients that only follow the last 10,000 or so blocks but for nodes fast syncing or state syncing this is impossible unless they download the versions and there are archive nodes still serving blocks of those versions

Reply:
Ah true, good point!